wrappers/golang/core/slice.go

```diff
@@ -0,0 +1,240 @@
+package core
+
+import (
+	"unsafe"
+
+	"github.com/ingonyama-zk/icicle/v3/wrappers/golang/runtime"
+)
+
+type HostOrDeviceSlice interface {
+	Len() int
+	Cap() int
+	IsEmpty() bool
+	IsOnDevice() bool
+	AsUnsafePointer() unsafe.Pointer
+}
+
+type DevicePointer = unsafe.Pointer
+
+type DeviceSlice struct {
+	inner unsafe.Pointer
+	// capacity is the number of bytes that have been allocated
+	capacity int
+	// length is the number of elements that have been written
+	length int
+}
+
+func (d DeviceSlice) Len() int {
+	return d.length
+}
+
+func (d DeviceSlice) Cap() int {
+	return d.capacity
+}
+
+func (d DeviceSlice) IsEmpty() bool {
+	return d.length == 0
+}
+
+func (d DeviceSlice) AsUnsafePointer() unsafe.Pointer {
+	return d.inner
+}
+
+func (d DeviceSlice) IsOnDevice() bool {
+	return true
+}
+
+// CheckDevice is used to ensure that the DeviceSlice about to be used resides on the currently set device
+func (d DeviceSlice) CheckDevice() {
+	if !runtime.IsActiveDeviceMemory(d.AsUnsafePointer()) {
+		panic("Attempt to use DeviceSlice on a different device")
+	}
+}
+
+func (d *DeviceSlice) Range(start, end int, endInclusive bool) DeviceSlice {
+	if end <= start {
+		panic("Cannot have negative or zero size slices")
+	}
+
+	if end >= d.length {
+		panic("Cannot increase slice size from Range")
+	}
+
+	var newSlice DeviceSlice
+	switch {
+	case start < 0:
+		panic("Negative value for start is not supported")
+	case start == 0:
+		newSlice = d.RangeTo(end, endInclusive)
+	case start > 0:
+		tempSlice := d.RangeFrom(start)
+		newSlice = tempSlice.RangeTo(end-start, endInclusive)
+	}
+	return newSlice
+}
+
+func (d *DeviceSlice) RangeTo(end int, inclusive bool) DeviceSlice {
+	if end <= 0 {
+		panic("Cannot have negative or zero size slices")
+	}
+
+	if end >= d.length {
```

Comment:
There is an off by one error here when `inclusive` is false
```suggestion
	if (inclusive && end >= d.length) || (!inclusive && end > d.length) {
```